src/python/pants/ox.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import contextlib
+import functools
+import importlib.machinery
+import logging
+import os
+import runpy
+import site
+import sys
+import zipimport
+
+_logger = logging.getLogger(__name__)
+
+
+def bootstrap_pyoxidizer() -> None:
+    if is_oxidized:
+        _logger.info("Pants is running as a PyOxidizer binary.")
+
+
+# Provide the `is_oxidized` symbol, to allow for workarounds in Pants code where we use things
+# that don't work under PyOxidizer's custom importer. `oxidized_importer` is only accessible
+# in Pants under PyOxidizer, so an import failure will occur if we're not oxidized.
+try:
+    import oxidized_importer  # type: ignore # pants: no-infer-dep # noqa: F401
+
+    is_oxidized = True
+except ModuleNotFoundError:
+    is_oxidized = False
+
+
+if is_oxidized and not sys.argv[0]:
+    # A not insignificant amount of Pants code relies on `sys.argv[0]`, which is modified in an
+    # invalid way by python's `pymain_run_module` support. For our purposes, the executable
+    # distribution is the correct `argv[0]`.
+    # See https://github.com/indygreg/PyOxidizer/issues/307
+    sys.argv[0] = sys.executable
+
+
+def pex_main() -> bool:
+    """Detect whether some external process is trying to invoke this binary as Pex.
+
+    When bootstrapping a plugin venv, Pants will reinvoke Pex in the binary that Pants was loaded
+    with. Pex will subsequently run a bunch of Python processes in order to run pip, venv, and
+    `python -c` invocations. This looks for certain command line switches, and attempt to invoke the
+    relevant modules that those switches indicate.
+    """
+
+    if len(sys.argv) < 2:
+        return False
+
+    if sys.argv[1] == "./pex":
```

Comment:
The "Detect whether some external process is trying to invoke this binary as Pex." bit seems like it might be a bit misleading. It's actually trying to use "this binary as _Python_", right? Because `argv[0]` is expected to be the python interpreter, and `argv[1]` is the argument to the interpreter.

Probably just a question of naming, but.

Reply:
These are all invocations that are Pants attempting to launch `pex`, or that are effected by `pex` attempting to bootstrap our plugins. The only machinery that is present is the machinery that supports `pex` (and then again, only for this extremely specific workflow), and I don't particularly want to give the impression that we're making a generic Python interpreter here.

Open to better names, but calling it "as Python" seems equally misleading

Reply:
(I changed the first line of the comment to strongly imply that we only support Pex use cases)